src/textual/widgets/_selection_list.py

```diff
@@ -0,0 +1,655 @@
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Callable, ClassVar, Generic, Iterable, TypeVar, cast
+
+from rich.repr import Result
+from rich.segment import Segment
+from rich.style import Style
+from rich.text import Text, TextType
+from typing_extensions import Self
+
+from ..binding import Binding
+from ..messages import Message
+from ..strip import Strip
+from ._option_list import NewOptionListContent, Option, OptionList
+from ._toggle_button import ToggleButton
+
+SelectionType = TypeVar("SelectionType")
+"""The type for the value of a [`Selection`][textual.widgets.selection_list.Selection] in a [`SelectionList`][textual.widgets.SelectionList]"""
+
+MessageSelectionType = TypeVar("MessageSelectionType")
+"""The type for the value of a [`Selection`][textual.widgets.selection_list.Selection] in a [`SelectionList`][textual.widgets.SelectionList] message."""
+
+
+class SelectionError(TypeError):
+    """Type of an error raised if a selection is badly-formed."""
+
+
+class Selection(Generic[SelectionType], Option):
+    """A selection for a [`SelectionList`][textual.widgets.SelectionList]."""
+
+    def __init__(
+        self,
+        prompt: TextType,
+        value: SelectionType,
+        initial_state: bool = False,
+        id: str | None = None,
+        disabled: bool = False,
+    ):
+        """Initialise the selection.
+
+        Args:
+            prompt: The prompt for the selection.
+            value: The value for the selection.
+            initial_state: The initial selected state of the selection.
+            id: The optional ID for the selection.
+            disabled: The initial enabled/disabled state. Enabled by default.
+        """
+        if isinstance(prompt, str):
+            prompt = Text.from_markup(prompt)
+        super().__init__(prompt.split()[0], id, disabled)
+        self._value: SelectionType = value
+        """The value associated with the selection."""
+        self._initial_state: bool = initial_state
+        """The initial selected state for the selection."""
+
+    @property
+    def value(self) -> SelectionType:
+        """The value for this selection."""
+        return self._value
+
+    @property
+    def initial_state(self) -> bool:
+        """The initial selected state for the selection."""
+        return self._initial_state
+
+
+class SelectionList(Generic[SelectionType], OptionList):
+    """A vertical selection list that allows making multiple selections."""
+
+    BINDINGS = [Binding("space", "select")]
+    """
+    | Key(s) | Description |
+    | :- | :- |
+    | space | Toggle the state of the highlighted selection. |
+    """
+
+    COMPONENT_CLASSES: ClassVar[set[str]] = {
+        "selection-list--button",
+        "selection-list--button-selected",
+        "selection-list--button-highlighted",
+        "selection-list--button-selected-highlighted",
+    }
+    """
```

Comment:
I'd suggest adding a note here to also check the component classes of `OptionList`. (Which would in turn mean you can remove that explicit note from the `SelectionList` docs page because it'd be inserted automatically.)

Reply:
The downside I saw to that was that it folk would have to go off on a sidequest to review those component classes; whereas with this approach they're all in the one document they're reading.